mlflow/pyspark/ml/__init__.py

```diff
@@ -218,18 +302,48 @@ def _log_pretraining_metadata(estimator, params):
         if params and isinstance(params, dict):
             estimator = estimator.copy(params)
 
+        param_map = _get_instance_param_map(estimator)
+
         # Chunk model parameters to avoid hitting the log_batch API limit
         for chunk in _chunk_dict(
-            _get_instance_param_map(estimator), chunk_size=MAX_PARAMS_TAGS_PER_BATCH,
+            param_map, chunk_size=MAX_PARAMS_TAGS_PER_BATCH,
         ):
             truncated = _truncate_dict(chunk, MAX_ENTITY_KEY_LENGTH, MAX_PARAM_VAL_LENGTH)
             try_mlflow_log(mlflow.log_params, truncated)
 
+        if _is_parameter_search_estimator(estimator):
+            tuning_param_maps = []
+            for eps in estimator.getEstimatorParamMaps():
+                tuning_param_maps.append({f'{k.parent}.{k.name}': str(v) for k, v in eps.items()})
+
+            try_mlflow_log(mlflow.log_dict, tuning_param_maps, 'estimatorParamMaps.json')
+
         try_mlflow_log(mlflow.set_tags, _get_estimator_info_tags(estimator))
 
     def _log_posttraining_metadata(estimator, spark_model, params):
-        # TODO: Log nested runs for spark ml tuning estimators
-        #   (CrossValidator/TrainValidationSplit)
+
+        if _is_parameter_search_estimator(estimator):
+            try:
+                # Fetch environment-specific tags (e.g., user and source) to ensure that lineage
+                # information is consistent with the parent run
+                child_tags = context_registry.resolve_tags()
+                child_tags.update({MLFLOW_AUTOLOGGING: AUTOLOGGING_INTEGRATION_NAME})
+                _create_child_runs_for_parameter_search(
+                    parent_estimator=estimator,
+                    parent_model=spark_model,
+                    parent_run=mlflow.active_run(),
+                    child_tags=child_tags
+                )
+            except Exception as e:
+                import traceback
+                msg = (
+                    "Encountered exception during creation of child runs for parameter search."
+                    " Child runs may be missing. Exception: {}".format(traceback.format_exc())
+                )
+                _logger.warning(msg)
+
+        # TODO: log parameter search results (best model parameters) as an artifact ?
```

Comment:
@harupy Thoughts ?

Reply:
If it helps, the scikit-learn autologging integration for GridSearch logs the best model parameters to the top-level (parent) run at the end of training. I think it would make sense to do the same here.

Reply:
In sklearn autologging, we log the `cv_results` attribute of a fitted parameter search estimator as an csv file in the parent run.

![Screen Shot 2021-04-21 at 11 58 12](https://user-images.githubusercontent.com/17039389/115490520-394d5f00-a299-11eb-86c3-6c337bb65721.png)

https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html

```
python examples/sklearn_autolog/grid_search_cv.py
mlflow ui
```

Reply:
^ For sklearn, `cv_results` contains all explored parameters and metrics, which is great to log as an artifact. I think we should do something similar for pyspark *and* log the parameters of the best estimator.

Reply:
btw does `CrossValidatorModel` store stats like above?

Reply:
We could create a table like below and log it as a csv file:

```
| param1 | param2 | avgMetric |
| :----- | :----- | :-------- |
| 1      | 0.1    | 0.4       |
| 1      | 0.2    | 0.5       |
| 2      | 0.1    | 0.6       |
| 2      | 0.2    | 0.7       |
```

Reply:
Currently the param name is `{estimator_uid}.{param_name}`, it include a `.`, seems cannot be used as csv column name. I use json format, does it make sense?

Reply:
> seems cannot be used as csv column name

Is this really true?